mountpoint-s3/src/inode/negative_cache.rs

```diff
@@ -0,0 +1,204 @@
+struct Key {
+    parent_ino: InodeNo,
+    child_name: String,
+}
+
+impl NegativeCache {
+    pub fn new(max_size: usize, ttl: Duration) -> Self {
+        Self {
+            map: RwLock::new(Default::default()),
+            max_size,
+            ttl,
+        }
+    }
+
+    /// Check whether the cache contains a **current** entry for the given
+    /// (`parent_ino`, `child_name`) pair.
+    pub fn contains(&self, parent_ino: InodeNo, child_name: &str) -> bool {
+        let key = Key {
+            parent_ino,
+            child_name: child_name.to_owned(),
+        };
+        let start = Instant::now();
+        let contains_current = self
+            .map
+            .read()
+            .unwrap()
+            .get(&key)
+            .is_some_and(|expiry| !expiry.is_expired());
+        metrics::histogram!(
+            "metadata_cache.negative_cache.operation_duration_us",
+            start.elapsed().as_micros() as f64,
+            "op" => "contains",
+        );
+        metrics::counter!("metadata_cache.negative_cache.cache_hit", contains_current.into());
+        contains_current
+    }
+
+    /// Remove an entry from the cache. If the entry was not present, this is a no-op.
+    pub fn remove(&self, parent_ino: InodeNo, child_name: &str) {
+        let key = Key {
+            parent_ino,
+            child_name: child_name.to_owned(),
+        };
+        let start = Instant::now();
+        let mut map = self.map.write().unwrap();
+        if map.remove(&key).is_some() {
+            metrics::gauge!("metadata_cache.negative_cache.entries", map.len() as f64);
+        }
+        metrics::histogram!(
+            "metadata_cache.negative_cache.operation_duration_us",
+            start.elapsed().as_micros() as f64,
+            "op" => "remove",
+        );
+    }
+
+    /// Insert an entry into the cache. If the entry already existed,
+    /// update its TTL.
+    /// Upon insertion, remove entries that exceed the cache limit or
+    /// that have already expired.
+    pub fn insert(&self, parent_ino: InodeNo, child_name: &str) {
+        let expiry = Expiry::from_now(self.ttl);
+        let key = Key {
+            parent_ino,
+            child_name: child_name.to_owned(),
+        };
+        let start = Instant::now();
+        let mut map = self.map.write().unwrap();
+        if map.insert(key, expiry).is_none() {
+            // Remove entries that have expired.
+            while map.front().is_some_and(|(_, e)| e.is_expired()) {
+                _ = map.pop_front();
+            }
+
+            // Remove entries that exceed the limit.
+            while map.len() > self.max_size {
+                let Some((_, e)) = map.pop_front() else {
+                    break;
+                };
+                // Report how many entries are evicted while still current.
+                metrics::counter!(
+                    "metadata_cache.negative_cache.entries_evicted_before_expiry",
```

Comment:
This metric will help detect cases where the negative cache is full. I don't currently see much value in monitoring the expirations. Besides, we track the current total, insertions, and deletions, so it could be derived.